### ISSUE TITLE
Investigate jerkiness in Earth orbit

### ISSUE THREAD

**Issue by ariel-phet** [phetsims/gravity-and-orbits#162]
currently the sim is running at high frame rate (60 fps on my computer) so it seems like the earth motion around the sun should be buttery smooth but it seems a bit jerky.

**Comment by aaronsamuel137** [comment]
Assigning to @ariel-phet for review. See http://www.colorado.edu/physics/phet/dev/html/gravity-and-orbits/1.0.0-dev.22/gravity-and-orbits_en.html

**Comment by ariel-phet** [comment]
discussed at dev meeting, sounds like we want to use a dt solution that was employed in Neuron.

**Comment by aaronsamuel137** [comment]
After looking at the clock with @jonathanolson, he said that it looks good to him. The issue of "choppiness" in FF is more likely a GC issue.

**Comment by aaronsamuel137** [comment]
Last I heard, @ariel-phet can still reproduce this on Win7 Chrome. I haven't been able to reproduce exactly what he sees, but there are still performance issues with GCs occasionally on FF.

I've done quite a bit to reduce allocations already, but there may need to be one last round of performance optimizations for this sim.

Assigning to @jessegreenberg since I won't have time to take care of this before I leave.

**Comment by ariel-phet** [comment]
Requested @samreid to document suggestions @jonathanolson had for improving performance

**Comment by samreid** [comment]
The suggestion I heard @jonathanolson describing was to use the canvas for rendering the paths, but without clearing the canvas each frame.  It would render new pixels at the head of the path, and draw over (erase) the pixels at the tail, but without clearing and re-rendering the entire path.

**Comment by samreid** [comment]
Also noting that PathsNode.js is already using CanvasNode, but not using deltas as described above.  I recommend using profiling tools or "binary search" to identify performance hot spots before working on the implementing the change as described above.

**Comment by jessegreenberg** [comment]
I would expect that changes in #207 should also help with this.  I noticed that the fading trace was actually  extending for about 1.5 orbits (in the default setting) while only 15% of that was visible.  In addition, invalidatePaint was being called when points were added and removed from the path.  The new method of drawing the trace from #207 only requires the path to be drawn when points are added.  We should now be calling invalidatePaint half as often.

**Comment by jessegreenberg** [comment]
Performance is great on iPad2 ad Windows 10 Chrome and Firefox.  I don't observe any jerkiness in an orbit on those platforms in any of the planetary systems in this sim.  As part of https://github.com/phetsims/tasks/issues/679 it would be good to know if this is still an issue.

**Comment by jessegreenberg** [comment]
While looking at the expensive frames, I noticed Color was taking up a bit of time.  A Color was being instantiated, but converted immediately to CSS.  The above commit uses a string directly.

**Comment by jessegreenberg** [comment]
At this point, it seems most noticeable for the space station orbit around the earth.

**Comment by jessegreenberg** [comment]
In the step function, new ModelStates and BodyStates are created frequently.  These are immutable representations of the model, and the bodies are updated from the values in these states.  But these are big copies of the body and main models.  They take up a pretty large chunk of the memory profile.  In Firefox, we see chops in the animation, which looks like garbage collection.  I think we will see better performance if we reference dynamic objects rather than create many static objects on every animation frame.

**Comment by jessegreenberg** [comment]
@jonathanolson and I met today to profile and discuss possible solutions for this.  The performance behaves as if the garbage collector is burdened, and the following memory profile (observed in Firefox) supports this:
![capture](https://cloud.githubusercontent.com/assets/6396244/18333196/34893b70-752a-11e6-8cb1-6ba826349d40.PNG)
The red bars are time spent by the garbage collector.

In the Chrome profile tools, we noticed a lot memory taken by compiled code, which indicated an excessive number of closures.  Indeed, we found function closures being created in PathsCanvasNode every time a body changes position.  Further searching found closures used by listeners attached to nearly every observable Property, including those that change every animation frame.

The above commits remove these closures.

**Comment by samreid** [comment]
What does the memory trace and performance look like after removing the closures?

**Comment by jessegreenberg** [comment]
Performance does look better!  Here is a pic of the memory trace after the above commits:
![capture](https://cloud.githubusercontent.com/assets/6396244/18333552/5a2babee-752d-11e6-9de2-80ae08de7778.PNG)

The number of GC's and the time they take have gone down. One thing that is odd here though is that there are more than one calls to requestAnimationFrame before paint is called.  In the picture above, requestAnimationFrame is called 5 times and it looks like paint is called only once.

**Comment by jessegreenberg** [comment]
The above pic does show some worse performing areas where the GC still takes up some time.  Around 3-4 seconds, we can see minor GC's lowering the fps.  Most GC's after that are CC graph reductions.  I guess these are faster and are actually preparingthe GC for cycle collections. https://developer.mozilla.org/en-US/docs/Tools/Performance/Waterfall#Markers

**Comment by jessegreenberg** [comment]
![capture](https://cloud.githubusercontent.com/assets/6396244/18357596/ce0a93d2-75ae-11e6-85cc-cc47d37b615b.PNG)

I think the remaining observable jerkiness is caused by the above.  The above image is a capture of a 1 second long garbage collection.  It happened within the first 6 seconds of running the orbit.  After this cycle collection, orbit animation was quite smooth for the remainder of runtime.

**Comment by samreid** [comment]
I am seeing the same long pause on OS X 10.11.6 and Firefox 48.0.1.

**Comment by samreid** [comment]
Launching fresh firefox and launching http://www.colorado.edu/physics/phet/dev/html/gravity-and-orbits/1.0.0-dev.30/gravity-and-orbits_en.html I did not see this problem at all in 2 trials.

**Comment by pixelzoom** [comment]
@ariel-phet asked me to take a peek.

**Comment by pixelzoom** [comment]
@ariel-phet asked me to take a peek. @jessegreenberg can you please point me to the relevant code?

**Comment by jessegreenberg** [comment]
There are the main places where we have been looking:

GravityAndOrbitsModel.step - computes the next state for each body based on the state of the whole orbital system
Body.addPathPoint - Adds a point to the body's path then triggers the canvas to repaint

**Comment by pixelzoom** [comment]
All orbits look generally jerky on Mac + FF 48.0.1.  And the jerkiness is chronic, not just for the first orbit.  So I don't think I can be of much help addressing the issue.

My test machine is a MacBookPro (5,3) with Mac OS X 10.11.6, with multiprocessing enabled in FF 48.0.1.  (Details on multiprocess FF at https://blog.mozilla.org/futurereleases/2016/08/02/whats-next-for-multi-process-firefox/.)

**Comment by samreid** [comment]
@jessegreenberg does the problem disappear if you reduce your screen resolution?

**Comment by samreid** [comment]
PathsCanvasNode.pointAddedListener is showing a lot of allocation:

![image](https://cloud.githubusercontent.com/assets/679486/18559949/c38b139a-7b36-11e6-80a2-caf0a5455a93.png)

**Comment by samreid** [comment]
@jessegreenberg can you comment out the body of pointAddedListener and see if the problem goes away.

**Comment by pixelzoom** [comment]
*Much* smoother on Mac+FF when the body of pointAddedListener is commented out. There's a little stutter on the first orbit, then looks good.

**Comment by samreid** [comment]
I tested this 5 times on Firefox 48.0.1 on Windows.  Twice I saw the hitch during animation.  The times I did not see the hitch during animation, it took slightly longer for the home screen to get to the sim screen, and the hitch may have happened during that step.

Even when I launch with ?screens=1 to avoid the homescreen and get to the path as soon as possible, there is only a hitch about half the time.

**Comment by jessegreenberg** [comment]
On Windows 10 Firefox 49.0 I saw the issue in 10/12 trials so I think I am seeing this more frequently than @samreid.

> Running with ?screens=1 doesn't show the problem (or very rarely)

For me, reducing the number of screens does not change the frequency at which this occurs.

> See how many canvases are used in the sim, using the scenery popupDebug facility

I am never really sure how to interpret popupDebug, but I think this shot shows that only one canvas is being used:
![display](https://cloud.githubusercontent.com/assets/6396244/18757023/53459c12-80af-11e6-89ce-573838cb3a1e.PNG)

> Use the strategy @kathy-phet proposed about running code in the home screen to try to trigger the initial GC sooner.

I tried to 'rev the engines' by adding this bit of code to CartoonScreen.  @samreid recommended to try packing into a setTimeout to be run asynchronously:
```
    setTimeout( function() {
      // enable animation
      cartoonModule.playButtonPressedProperty.set( true );
      for ( var i = 0; i < 5000; i++ ) {
        cartoonModule.step( 0.16 );
      }
      // reset model for user use and reset playing property
      cartoonModule.reset();
    }, 200 );
```
This didn't assist with the issue, but I will verify that this is really generating enough garbage.

**Comment by jessegreenberg** [comment]
I tried a *very quick* 'hello world' kind of test using WebGL, removing all bodies and replacing the earth with a triangle.
![webgl](https://cloud.githubusercontent.com/assets/6396244/18800027/0f1a4ada-8198-11e6-8632-d63c2d3a89e9.PNG)

Even in this case, with fewer bodies being drawn with WebGL, I see the hiccup frequently when the orbit starts.  I will check the code into a branch just in case we want to revisit using WebGL later, but this really is a model/GC issue and view optimizations won't help.

**Comment by jessegreenberg** [comment]
@jonathanolson provided a link to additional Firefox memory profiling tools that are not enabled by default:
https://developer.mozilla.org/en-US/docs/Tools/Memory/Basic_operations#Opening_the_Memory_tool

This will let us see heap snapshots, compare by diffing snapshots, and even record the call stack.

**Comment by samreid** [comment]
> In the step function, new ModelStates and BodyStates are created frequently. These are immutable representations of the model, and the bodies are updated from the values in these states. But these are big copies of the body and main models. They take up a pretty large chunk of the memory profile. In Firefox, we see chops in the animation, which looks like garbage collection. I think we will see better performance if we reference dynamic objects rather than create many static objects on every animation frame.

Have you tried changing the model code so that there are fewer allocations? Using pooling or mutable physics?  A quick hack that will tell you whether we are barking up the right tree would be to change toBodyState/new BodyState() and all of those copy() functions and see what the waterfall view looks like.  For instance, you could remove the copy functions in toBodyState and record the waterfall view and see if there is a long GC even though the physics would be wrong.  If the waterfall looks good, then you could investigate pooling or rewriting the physics to avoid too many allocations.

**Comment by jonathanolson** [comment]
Just did some quick testing, and couldn't reproduce a major stall (only had some very minor one once). Things seemed smooth.

FF's memory tools shows that modelToViewPosition's usage in pointAddedListener and the toBodyState copy were the main allocations.

**Comment by jessegreenberg** [comment]
> Have you tried changing the model code so that there are fewer allocations?

Yes I did, but it did not improve the issue at all.

> you could remove the copy functions in toBodyState and record the waterfall view and see if there is a long GC even though the physics would be wrong.

Good idea, I gave this a try and the issue persists with the same frequency.

Going down this path, I continued to comment out places in the sim where things were getting allocated frequently.  No matter what, the major stall was observable using the profiling tools.  I *completely comment out step* in GravityAndOrbitsModule.js, and the major glitch was still observable in the Firefox profile.  This indicated that it has nothing to do with the initial orbit, but rather the sim startup.

In b702067de289f095ca25bac028457881db7bfe44 (branch hack-n-slash) I created a new testing screen view, with a simplified model that only had a model and view components for the earth + sun system.  I removed all of the systems. In this simplified setup, major jerkiness was still observed without animation in the Firefox profile.

To narrow it down, I tried starting from scratch, replacing the model and view with the following in TestScreen, the *only* Screen being used in the sim:
```
    Screen.call( this,
      function() { return {} },
      function( model ) { return new ScreenView( model ); },
      options
    );
```

The problem is still present for both RequireJS and Built versions!  Here is an example of the memory profile in this case:
![nomodelorview](https://cloud.githubusercontent.com/assets/6396244/18926540/6e781602-8575-11e6-9d20-3b837a48b6e8.PNG)

I am wondering if this is happening in other sims, but they generally don't start animation so quickly after startup so it isn't noticeable?

**Comment by jessegreenberg** [comment]
Or maybe Firefox is caching the optimization, applying it aggressively in my tests above? I don't know how that would work, and it seems unlikely because I was able to reproduce the above image several times.

**Comment by jessegreenberg** [comment]
I just observed this in the first screen of forces-and-motion-basics (although with less frequency).  Here is an example, I saw this in 3 / 9 trials:

![famb-1](https://cloud.githubusercontent.com/assets/6396244/18926973/102d57b8-8577-11e6-9b95-86cf0674bb84.PNG)

The above shows an 800ms delay for garbage collection.

**Comment by jessegreenberg** [comment]
Based on https://github.com/phetsims/gravity-and-orbits/issues/162#issuecomment-250254257 and https://github.com/phetsims/gravity-and-orbits/issues/162#issuecomment-250257015, it seems like we should move this issue to scenery, and/or monitor Firefox to see if this improves in betas or later browser versions.

@jonathanolson and @ariel-phet what do you think?

**Comment by ariel-phet** [comment]
@jessegreenberg I think you are right. Please move to scenery and mark as a browser bug.

Looks like something we will have to live with for the moment (fortunately our usage is mostly dominated by Chrome these days)

**Comment by jessegreenberg** [comment]
Issue has been created in scenery.  It will be tracked there.  Removing the high priority label for the moment, and marking as deferred as mentioned in https://github.com/phetsims/gravity-and-orbits/issues/162#issuecomment-250321585.

**Comment by samreid** [comment]
The October 2019 test reported in https://github.com/phetsims/qa/issues/443#issuecomment-545496827 did not indicate significant problems, closing.
